runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -101,19 +119,64 @@ public SimpleDoFnRunner(
       TupleTag<OutputT> mainOutputTag,
       List<TupleTag<?>> additionalOutputTags,
       StepContext stepContext,
+      @Nullable Coder<InputT> inputCoder,
+      Map<TupleTag<?>, Coder<?>> outputCoders,
       WindowingStrategy<?, ?> windowingStrategy) {
     this.options = options;
     this.fn = fn;
     this.signature = DoFnSignatures.getSignature(fn.getClass());
     this.observesWindow = signature.processElement().observesWindow() || !sideInputReader.isEmpty();
     this.invoker = DoFnInvokers.invokerFor(fn);
     this.sideInputReader = sideInputReader;
+    this.schemaCoder = (inputCoder != null && inputCoder instanceof SchemaCoder)
+        ? (SchemaCoder<InputT>) inputCoder : null;
+    this.outputCoders = outputCoders;
+    if (outputCoders != null) {
+      Coder<OutputT> outputCoder = (Coder<OutputT>) outputCoders.get(mainOutputTag);
+      mainOutputSchemaCoder = (outputCoder instanceof SchemaCoder)
+          ? (SchemaCoder<OutputT>) outputCoder : null;
+    } else {
+      mainOutputSchemaCoder = null;
+    }
     this.outputManager = outputManager;
     this.mainOutputTag = mainOutputTag;
     this.outputTags =
         Sets.newHashSet(FluentIterable.<TupleTag<?>>of(mainOutputTag).append(additionalOutputTags));
     this.stepContext = stepContext;
 
+    // Currently we only support a single FieldAccess on a processElement. We should decide
+    // whether to get rid of the FieldAccess ids, or find a use for multiple.
+    DoFnSignature doFnSignature = DoFnSignatures.getSignature(fn.getClass());
+    DoFnSignature.ProcessElementMethod processElementMethod = doFnSignature.processElement();
+    RowParameter rowParameter = processElementMethod.getRowParameter();
+    FieldAccessDescriptor fieldAccessDescriptor = null;
+    if (rowParameter != null) {
+      checkArgument(schemaCoder != null,
+          "Cannot access object as a row if the input PCollection does not have a schema ."
```

Comment:
Does "no schema coder" ==> "no schema" or do we have to talk about not having the schema coder defined in this exception message ?

Reply:
Schema is currently represented under the covers by SchemaCoder. So no schema coder === no schema.

Reply:
Schema is currently represented under the covers by SchemaCoder. So no schema coder === no schema.